--- a/libc/sysdeps/linux/common/Makefile.in
+++ b/libc/sysdeps/linux/common/Makefile.in
@@ -24,6 +24,7 @@
 	capget.c \
 	capset.c \
 	eventfd.c \
+	fanotify.c \
 	inotify.c \
 	ioperm.c \
 	iopl.c \
--- a/dev/null
+++ b/libc/sysdeps/linux/common/fanotify.c
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2016, The Linux Foundation. All rights reserved.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+
+
+#include <sys/syscall.h>
+#include <sys/fanotify.h>
+
+#ifdef __NR_fanotify_init
+_syscall2(int, fanotify_init, unsigned int, flags, unsigned int, event_f_flags)
+#endif
+
+#ifdef __NR_fanotify_mark
+#include <bits/wordsize.h>
+#include <fcntl.h>
+
+#if __WORDSIZE == 64
+_syscall5(int, fanotify_mark, int, fanotify_fd, unsigned int, flags,
+	uint64_t, mask, int, dirfd, const char *, pathname)
+#else
+int fanotify_mark(int fanotify_fd, unsigned int flags,
+		uint64_t mask, int dirfd, const char *pathname)
+{
+	return INLINE_SYSCALL(fanotify_mark, 6, fanotify_fd, flags,
+				__LONG_LONG_PAIR(mask >> 31,mask), dirfd, pathname);
+}
+
+#endif
+#endif
+
+
+
--- a/dev/null
+++ b/libc/sysdeps/linux/common/sys/fanotify.h
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2016, The Linux Foundation. All rights reserved.
+ *
+ * Permission to use, copy, modify, and/or distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+
+
+#ifndef	_SYS_FANOTIFY_H
+#define	_SYS_FANOTIFY_H	1
+
+#include <stdint.h>
+
+__BEGIN_DECLS
+/* Create and initialize fanotify group.  */
+extern int fanotify_init (unsigned int __flags, unsigned int __event_f_flags)
+__THROW;
+
+/* Add, remove, or modify an fanotify mark on a filesystem object.  */
+extern int fanotify_mark (int __fanotify_fd, unsigned int __flags,
+				uint64_t __mask, int __dfd, const char *__pathname)
+__THROW;
+__END_DECLS
+
+#endif
