Index: netifd-2015-09-10/bridge.c
===================================================================
--- netifd-2015-09-10.orig/bridge.c	2015-10-15 22:19:07.000000000 +0530
+++ netifd-2015-09-10/bridge.c	2015-10-17 15:42:33.658497665 +0530
@@ -32,6 +32,7 @@
 	BRIDGE_ATTR_HELLO_TIME,
 	BRIDGE_ATTR_MAX_AGE,
 	BRIDGE_ATTR_BRIDGE_EMPTY,
+	BRIDGE_ATTR_HAIRPIN,
 	BRIDGE_ATTR_MULTICAST_QUERIER,
 	BRIDGE_ATTR_HASH_MAX,
 	__BRIDGE_ATTR_MAX
@@ -47,6 +48,7 @@
 	[BRIDGE_ATTR_MAX_AGE] = { "max_age", BLOBMSG_TYPE_INT32 },
 	[BRIDGE_ATTR_IGMP_SNOOP] = { "igmp_snooping", BLOBMSG_TYPE_BOOL },
 	[BRIDGE_ATTR_BRIDGE_EMPTY] = { "bridge_empty", BLOBMSG_TYPE_BOOL },
+	[BRIDGE_ATTR_HAIRPIN] = { "hairpin", BLOBMSG_TYPE_BOOL },
 	[BRIDGE_ATTR_MULTICAST_QUERIER] = { "multicast_querier", BLOBMSG_TYPE_BOOL },
 	[BRIDGE_ATTR_HASH_MAX] = { "hash_max", BLOBMSG_TYPE_INT32 },
 };
@@ -207,6 +209,7 @@
 	}
 
 	device_set_present(&bst->dev, true);
+	system_bridge_set_hairpin(bm->dev.dev, bst->config.hairpin);
 	device_broadcast_event(&bst->dev, DEV_EVENT_TOPO_CHANGE);
 
 	return 0;
@@ -556,6 +559,7 @@
 	cfg->hash_max = 512;
 	cfg->bridge_empty = false;
 	cfg->priority = 0x7FFF;
+	cfg->hairpin = false;
 
 	if ((cur = tb[BRIDGE_ATTR_STP]))
 		cfg->stp = blobmsg_get_bool(cur);
@@ -592,6 +596,9 @@
 
 	if ((cur = tb[BRIDGE_ATTR_BRIDGE_EMPTY]))
 		cfg->bridge_empty = blobmsg_get_bool(cur);
+
+	if ((cur = tb[BRIDGE_ATTR_HAIRPIN]))
+		cfg->hairpin = blobmsg_get_bool(cur);
 }
 
 enum dev_change_type
Index: netifd-2015-09-10/system-linux.c
===================================================================
--- netifd-2015-09-10.orig/system-linux.c	2015-10-17 15:26:35.741233181 +0530
+++ netifd-2015-09-10/system-linux.c	2015-10-17 15:45:14.003370870 +0530
@@ -686,6 +686,13 @@
 	return system_bridge_if(bridge->ifname, dev, SIOCBRDELIF, NULL);
 }
 
+void system_bridge_set_hairpin(struct device *dev, int enable)
+{
+	char mode[4];
+	snprintf(mode, sizeof(mode), "%d", !!enable);
+	system_set_dev_sysctl("/sys/class/net/%s/brport/hairpin_mode", dev->ifname, mode);
+}
+
 int system_if_resolve(struct device *dev)
 {
 	struct ifreq ifr;
Index: netifd-2015-09-10/system.h
===================================================================
--- netifd-2015-09-10.orig/system.h	2015-10-17 15:26:35.745233199 +0530
+++ netifd-2015-09-10/system.h	2015-10-17 15:46:45.723872353 +0530
@@ -58,6 +58,7 @@
 	int ageing_time;
 	int hello_time;
 	int max_age;
+	int hairpin;
 	int hash_max;
 };
 
@@ -109,6 +110,7 @@
 int system_bridge_delbr(struct device *bridge);
 int system_bridge_addif(struct device *bridge, struct device *dev);
 int system_bridge_delif(struct device *bridge, struct device *dev);
+void system_bridge_set_hairpin(struct device *dev, int enable);
 
 int system_bonding_addbonding(struct device *bonding, struct bonding_config *cfg);
 int system_bonding_delbonding(struct device *bonding);
