Index: uci-g424292.1/file.c
===================================================================
--- uci-g424292.1.orig/file.c	2016-04-07 10:59:32.366990682 +0530
+++ uci-g424292.1/file.c	2016-04-07 11:01:42.050991638 +0530
@@ -157,7 +157,6 @@
 			break;
 		}
 	}
-	uci_parse_error(ctx, "unterminated \"");
 }
 
 /*
@@ -180,14 +179,13 @@
 			/* Multi-line str value */
 			uci_getln(ctx, pctx->pos);
 			if (!pctx_cur_char(pctx)) {
-				uci_parse_error(ctx, "EOF with unterminated \"");
+				uci_parse_error(ctx, "EOF with unterminated '");
 			}
 			break;
 		default:
 			addc(ctx, target, &pctx->pos);
 		}
 	}
-	uci_parse_error(ctx, "unterminated '");
 }
 
 /*
