From b73e3f4a4beaeff1f050228afcd10bc6b43eccfd Mon Sep 17 00:00:00 2001
From: Sandhya Ks <sandhyks@codeaurora.org>
Date: Fri, 17 Jul 2020 11:32:35 +0530
Subject: fstools : Empty ubi volume if overlay mount fails after jffs2reset

Signed-off-by: Sandhya Ks <sandhyks@codeaurora.org>
---
 libfstools/overlay.c | 32 +++++++++++++++++++++++++++++---
 1 file changed, 29 insertions(+), 3 deletions(-)

diff --git a/libfstools/overlay.c b/libfstools/overlay.c
index 14214a3..91f096e 100644
--- a/libfstools/overlay.c
+++ b/libfstools/overlay.c
@@ -335,12 +335,14 @@ jffs2_switch(struct volume *v)
 static int overlay_mount_fs(struct volume *v)
 {
 	char *fstype = overlay_fs_name(volume_identify(v));
+	char cmd[64];
 
 	if (mkdir("/tmp/overlay", 0755)) {
 		ULOG_ERR("failed to mkdir /tmp/overlay: %m\n");
 		return -1;
 	}
 
+	ULOG_INFO("Mounting %s %s on /tmp/overlay\n",fstype,v->blk);
 	if (mount(v->blk, "/tmp/overlay", fstype,
 #ifdef OVL_MOUNT_FULL_ACCESS_TIME
 		MS_RELATIME,
@@ -353,9 +355,33 @@ static int overlay_mount_fs(struct volume *v)
 		NULL
 #endif
 		)) {
-		ULOG_ERR("failed to mount -t %s %s /tmp/overlay: %m\n",
-		         fstype, v->blk);
-		return -1;
+		if(!strncmp(fstype,"ubifs",strlen("ubifs"))) {
+			snprintf(cmd,sizeof(cmd),"ubiupdatevol -t %s",v->blk);
+			ULOG_ERR("Retrying after emptying ubi volume %s\n",v->blk);
+			ULOG_ERR("%s\n" ,cmd);
+			system(cmd);
+			if (mount(v->blk, "/tmp/overlay", fstype,
+#ifdef OVL_MOUNT_FULL_ACCESS_TIME
+			MS_RELATIME,
+#else
+			MS_NOATIME,
+#endif
+#ifdef OVL_MOUNT_COMPRESS_ZLIB
+			"compr=zlib"
+#else
+			NULL
+#endif
+			)) {
+				ULOG_ERR("failed to mount -t %s %s /tmp/overlay: %m\n",
+				 fstype, v->blk);
+				return -1;
+			}
+		}
+		else {
+			ULOG_ERR("failed to mount -t %s %s /tmp/overlay: %m\n",
+			 fstype, v->blk);
+			return -1;
+		}
 	}
 
 	return 0;
-- 
2.7.4

