Index: linux-3.3.8/drivers/net/ppp/pppoe.c
===================================================================
--- linux-3.3.8.orig/drivers/net/ppp/pppoe.c	2014-01-14 12:55:22.230480483 -0600
+++ linux-3.3.8/drivers/net/ppp/pppoe.c	2014-01-14 19:43:21.268046378 -0600
@@ -558,6 +558,9 @@
 	return 0;
 }
 
+struct pppox_sock *last_pppox_sock = NULL;
+EXPORT_SYMBOL(last_pppox_sock);
+
 static int pppoe_release(struct socket *sock)
 {
 	struct sock *sk = sock->sk;
@@ -595,6 +598,9 @@
 	 */
 	delete_item(pn, po->pppoe_pa.sid, po->pppoe_pa.remote,
 		    po->pppoe_ifindex);
+	if (po == last_pppox_sock) {
+		last_pppox_sock = NULL;
+	}
 
 	sock_orphan(sk);
 	sock->sk = NULL;
@@ -696,6 +702,7 @@
 	}
 
 	po->num = sp->sa_addr.pppoe.sid;
+	last_pppox_sock = po;
 
 end:
 	release_sock(sk);
Index: linux-3.3.8/include/linux/if_pppox.h
===================================================================
--- linux-3.3.8.orig/include/linux/if_pppox.h	2014-01-14 12:54:48.570480307 -0600
+++ linux-3.3.8/include/linux/if_pppox.h	2014-01-14 12:56:09.230479131 -0600
@@ -220,6 +220,8 @@
     PPPOX_DEAD		= 16  /* dead, useless, please clean me up!*/
 };
 
+extern struct pppox_sock *last_pppox_sock;
+
 #endif /* __KERNEL__ */
 
 #endif /* !(__LINUX_IF_PPPOX_H) */
