Index: hostapd-2016-10-02/src/ap/wps_hostapd.c
===================================================================
--- hostapd-2016-10-02.orig/src/ap/wps_hostapd.c	2017-04-24 13:32:28.861743089 +0530
+++ hostapd-2016-10-02/src/ap/wps_hostapd.c	2017-04-24 13:35:25.636422955 +0530
@@ -1067,14 +1067,11 @@
 		if (conf->rsn_pairwise & (WPA_CIPHER_CCMP | WPA_CIPHER_GCMP)) {
 			wps->encr_types |= WPS_ENCR_AES;
 			wps->encr_types_rsn |= WPS_ENCR_AES;
-		}
-		if (conf->rsn_pairwise & WPA_CIPHER_TKIP) {
+		} else if (conf->rsn_pairwise & WPA_CIPHER_TKIP) {
 			wps->encr_types |= WPS_ENCR_TKIP;
 			wps->encr_types_rsn |= WPS_ENCR_TKIP;
 		}
-	}
-
-	if (conf->wpa & WPA_PROTO_WPA) {
+	} else if (conf->wpa & WPA_PROTO_WPA) {
 		if (conf->wpa_key_mgmt & WPA_KEY_MGMT_PSK)
 			wps->auth_types |= WPS_AUTH_WPAPSK;
 		if (conf->wpa_key_mgmt & WPA_KEY_MGMT_IEEE8021X)
@@ -1083,8 +1080,7 @@
 		if (conf->wpa_pairwise & WPA_CIPHER_CCMP) {
 			wps->encr_types |= WPS_ENCR_AES;
 			wps->encr_types_wpa |= WPS_ENCR_AES;
-		}
-		if (conf->wpa_pairwise & WPA_CIPHER_TKIP) {
+		} else if (conf->wpa_pairwise & WPA_CIPHER_TKIP) {
 			wps->encr_types |= WPS_ENCR_TKIP;
 			wps->encr_types_wpa |= WPS_ENCR_TKIP;
 		}
