--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -117,9 +117,11 @@ static void hostapd_reload_bss(struct hostapd_data *hapd)
 		if (hostapd_set_generic_elem(hapd, wpa_ie, wpa_ie_len))
 			wpa_printf(MSG_ERROR, "Failed to configure WPA IE for "
 				   "the kernel driver.");
-	} else if (hapd->wpa_auth) {
-		wpa_deinit(hapd->wpa_auth);
-		hapd->wpa_auth = NULL;
+	} else {
+		if (hapd->wpa_auth) {
+			wpa_deinit(hapd->wpa_auth);
+			hapd->wpa_auth = NULL;
+		}
 		hostapd_set_privacy(hapd, 0);
 		hostapd_setup_encryption(hapd->conf->iface, hapd);
 		hostapd_set_generic_elem(hapd, (u8 *) "", 0);
