--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -3874,6 +3874,7 @@ static int hostapd_config_fill(struct ho
 	} else if (os_strcmp(buf, "sae_commit_override") == 0) {
 		wpabuf_free(bss->sae_commit_override);
 		bss->sae_commit_override = wpabuf_parse_bin(pos);
+#endif /* CONFIG_TESTING_OPTIONS */
 #ifdef CONFIG_SAE
 	} else if (os_strcmp(buf, "sae_password") == 0) {
 		if (parse_sae_password(bss, pos) < 0) {
@@ -3882,7 +3883,6 @@ static int hostapd_config_fill(struct ho
 			return 1;
 		}
 #endif /* CONFIG_SAE */
-#endif /* CONFIG_TESTING_OPTIONS */
 	} else if (os_strcmp(buf, "vendor_elements") == 0) {
 		if (parse_wpabuf_hex(line, buf, &bss->vendor_elements, pos))
 			return 1;
