Index: hostapd-2018-06-06/hostapd/config_file.c
===================================================================
--- hostapd-2018-06-06.orig/hostapd/config_file.c	2018-08-23 19:16:24.805276906 +0530
+++ hostapd-2018-06-06/hostapd/config_file.c	2018-08-23 19:16:24.801276906 +0530
@@ -1237,6 +1237,9 @@
 		conf->ht_capab |= HT_CAP_INFO_SUPP_CHANNEL_WIDTH_SET;
 		conf->ht40_plus_minus_allowed = 1;
 	}
+	if (!os_strstr(capab, "[HT40+]") && !os_strstr(capab, "[HT40-]")) {
+		conf->secondary_channel = 0;
+	}
 	if (os_strstr(capab, "[SMPS-STATIC]")) {
 		conf->ht_capab &= ~HT_CAP_INFO_SMPS_MASK;
 		conf->ht_capab |= HT_CAP_INFO_SMPS_STATIC;
