--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -3354,6 +3354,27 @@ static int hostapd_fill_csa_settings(str
 		return -1;
 
 	switch (settings->freq_params.bandwidth) {
+	case 40:
+		/* When changing to bw of 40Mhz, secondary channel offset flag should
+		* be set and flag in HT capabilities should be set to support bw of
+		* 40MHz
+		*/
+		if (settings->freq_params.center_freq1 < settings->freq_params.freq)
+			hapd->cs_freq_params.sec_channel_offset = -1;
+		else
+			hapd->cs_freq_params.sec_channel_offset = 1;
+
+		/* If secondary channel offset is not provided as parameter, it will
+		 * default to 0. If bw is 40, secondary channel offset needs to be
+		 * set to -1/1 to determine if center freq PLUS or MINUS
+		*/
+		if (!settings->freq_params.sec_channel_offset)
+			settings->freq_params.sec_channel_offset =
+			  hapd->cs_freq_params.sec_channel_offset;
+
+		hapd->iconf->ht_capab |= HT_CAP_INFO_SUPP_CHANNEL_WIDTH_SET;
+		bandwidth = CHANWIDTH_USE_HT;
+		break;
 	case 80:
 		if (settings->freq_params.center_freq2)
 			bandwidth = CHANWIDTH_80P80MHZ;
