Index: util-linux-2.21.1/fdisk/fdiskbsdlabel.h
===================================================================
--- util-linux-2.21.1.orig/fdisk/fdiskbsdlabel.h	2012-05-25 02:07:50.064972749 -0700
+++ util-linux-2.21.1/fdisk/fdiskbsdlabel.h	2012-05-25 02:07:19.100973421 -0700
@@ -48,7 +48,8 @@
 
 #if defined (__i386__) || defined (__sparc__) || defined (__arm__) || \
     defined (__mips__) || defined (__s390__) || defined (__sh__) || \
-    defined(__x86_64__) || defined (__avr32__) || defined(__cris__)
+    defined(__x86_64__) || defined (__avr32__) || defined(__cris__) || \
+    defined(__ubicom32__)
 #define BSD_LABELSECTOR   1
 #define BSD_LABELOFFSET   0
 #elif defined (__alpha__) || defined (__powerpc__) || defined (__ia64__) || defined (__hppa__)
