--- a/arch/mips/ath79/Kconfig
+++ b/arch/mips/ath79/Kconfig
@@ -114,6 +114,9 @@ config ATH79_DEV_AP9X_PCI
 	select ATH79_PCI_ATH9K_FIXUP
 	def_bool n
 
+config ATH79_DEV_AUDIO
+	def_bool n
+
 config ATH79_DEV_DSA
 	def_bool n
 
@@ -133,6 +136,9 @@ config ATH79_DEV_NFC
 	depends on (SOC_AR934X)
 	def_bool n
 
+config ATH79_DEV_NAND
+	def_bool n
+
 config ATH79_DEV_SPI
 	def_bool n
 
--- a/arch/mips/ath79/Makefile
+++ b/arch/mips/ath79/Makefile
@@ -18,11 +18,13 @@ obj-$(CONFIG_PCI)			+= pci.o
 #
 obj-y					+= dev-common.o
 obj-$(CONFIG_ATH79_DEV_AP9X_PCI)	+= dev-ap9x-pci.o
+obj-$(CONFIG_ATH79_DEV_AUDIO)		+= dev-audio.o
 obj-$(CONFIG_ATH79_DEV_DSA)		+= dev-dsa.o
 obj-$(CONFIG_ATH79_DEV_ETH)		+= dev-eth.o
 obj-$(CONFIG_ATH79_DEV_GPIO_BUTTONS)	+= dev-gpio-buttons.o
 obj-$(CONFIG_ATH79_DEV_LEDS_GPIO)	+= dev-leds-gpio.o
 obj-$(CONFIG_ATH79_DEV_M25P80)		+= dev-m25p80.o
+obj-$(CONFIG_ATH79_DEV_NAND)		+= dev-nand.o
 obj-$(CONFIG_ATH79_DEV_NFC)		+= dev-nfc.o
 obj-$(CONFIG_ATH79_DEV_SPI)		+= dev-spi.o
 obj-$(CONFIG_ATH79_DEV_USB)		+= dev-usb.o
