diff --git a/drivers/watchdog/ath79_wdt.c b/drivers/watchdog/ath79_wdt.c
index 1b57349..dac3815 100644
--- a/drivers/watchdog/ath79_wdt.c
+++ b/drivers/watchdog/ath79_wdt.c
@@ -113,6 +113,10 @@ static irqreturn_t ath79_wdt_irq_handler(int irq, void *dev_id)
 	pr_info("Watchdog bark! Now = %llu \n", t);
 	pr_info("Watchdog last pet at %llu \n", last_pet);
 	pr_info("\n ================================== \n");
+
+#if defined(CONFIG_KEXEC) && defined(CONFIG_FTRACE)
+	ftrace_dump(DUMP_ALL);
+#endif
 	panic("BUG :  ATH_WDT_TIMEOUT ");
 
 	return IRQ_HANDLED;
