--- a/ocaml/no_prepare_ocamlcocci.ml
+++ b/ocaml/no_prepare_ocamlcocci.ml
@@ -9,9 +9,9 @@
 	function
 	    Ast_cocci.ScriptRule (name,"ocaml",deps,mv,script_vars,code) ->
 	      code :: prev
-	  | Ast_cocci.InitialScriptRule (name,"ocaml",deps,mvs,code) ->
+	  | Ast_cocci.InitialScriptRule (name,"ocaml",deps,mv,code) ->
 	      code :: prev
-	  | Ast_cocci.FinalScriptRule (name,"ocaml",deps,mvs,code) ->
+	  | Ast_cocci.FinalScriptRule (name,"ocaml",deps,mv,code) ->
 	      code :: prev
 	  | _ -> prev)
       [] code in
