Index: uClibc-0.9.32/libc/sysdeps/linux/common/Makefile.in
===================================================================
--- uClibc-0.9.32.orig/libc/sysdeps/linux/common/Makefile.in	2012-05-25 13:34:26.830828706 -0700
+++ uClibc-0.9.32/libc/sysdeps/linux/common/Makefile.in	2012-05-25 13:34:19.014828856 -0700
@@ -24,7 +24,7 @@
 	remap_file_pages.c sched_getaffinity.c sched_setaffinity.c \
 	sendfile64.c sendfile.c setfsgid.c setfsuid.c setresuid.c \
 	splice.c vmsplice.c tee.c signalfd.c swapoff.c swapon.c \
-	sync_file_range.c sysctl.c sysinfo.c timerfd.c uselib.c vhangup.c
+	sync_file_range.c sysctl.c sysinfo.c timerfd.c uselib.c vhangup.c unshare.c
 # NPTL needs these internally: madvise.c
 CSRC-$(findstring y,$(UCLIBC_LINUX_SPECIFIC)$(UCLIBC_HAS_THREADS_NATIVE)) += madvise.c
 ifeq ($(UCLIBC_HAS_THREADS_NATIVE),y)
Index: uClibc-0.9.32/libc/sysdeps/linux/common/bits/sched.h
===================================================================
--- uClibc-0.9.32.orig/libc/sysdeps/linux/common/bits/sched.h	2012-05-25 13:34:26.614828710 -0700
+++ uClibc-0.9.32/libc/sysdeps/linux/common/bits/sched.h	2012-05-25 10:59:41.439029479 -0700
@@ -74,11 +74,9 @@
 extern int clone (int (*__fn) (void *__arg), void *__child_stack,
 		  int __flags, void *__arg, ...) __THROW;
 
-#if 0
 /* Unshare the specified resources.  */
 extern int unshare (int __flags) __THROW;
 #endif
-#endif
 
 __END_DECLS
 
Index: uClibc-0.9.32/libc/sysdeps/linux/common/stubs.c
===================================================================
--- uClibc-0.9.32.orig/libc/sysdeps/linux/common/stubs.c	2012-05-25 13:34:26.454828713 -0700
+++ uClibc-0.9.32/libc/sysdeps/linux/common/stubs.c	2012-05-25 11:02:13.951026238 -0700
@@ -278,6 +278,10 @@
 make_stub(umount2)
 #endif
 
+#if !defined __NR_unshare && defined __UCLIBC_LINUX_SPECIFIC__
+make_stub(unshare)
+#endif
+
 #ifndef __NR_utimensat
 make_stub(futimens)
 make_stub(utimensat)
Index: uClibc-0.9.32/libc/sysdeps/linux/common/unshare.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ uClibc-0.9.32/libc/sysdeps/linux/common/unshare.c	2012-05-15 00:20:09.000000000 -0700
@@ -0,0 +1,15 @@
+/* vi: set sw=4 ts=4: */
+/*
+ * unshare() for uClibc
+ *
+ * Copyright (C) 2011 Henning Heinold <heinold@inf.fu-berlin.de>
+ *
+ * Licensed under the LGPL v2.1, see the file COPYING.LIB in this tarball.
+ */
+
+#include <sys/syscall.h>
+#include <sched.h>
+
+#if defined __NR_unshare
+_syscall1(int, unshare, int, flags)
+#endif
